pkg/emitter/nats_data.go

```diff
@@ -0,0 +1,52 @@
+//
+// Copyright 2022 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package emitter
+
+import (
+	"context"
+	"time"
+)
+
+type DataFunc func([]byte) error
+
+type PubSub struct {
+	dataChan <-chan []byte
+	errChan  <-chan error
+}
+
+func NewProcessorSubscriber(ctx context.Context, id string, subj string, durable string, backOffTimer time.Duration) (*PubSub, error) {
+	dataChan, errchan, err := createSubscriber(ctx, id, subj, durable, backOffTimer)
+	if err != nil {
+		return nil, err
+	}
+	return &PubSub{
+		dataChan: dataChan,
+		errChan:  errchan,
+	}, nil
+}
+
+func (psub *PubSub) GetProcessorDataFromNats(ctx context.Context, processFunc DataFunc) error {
+	for {
+		select {
+		case d := <-psub.dataChan:
+			if err := processFunc(d); err != nil {
+				return err
+			}
+		case err := <-psub.errChan:
+			return err
+		}
+	}
```

Comment:
I think there’s an edge case here, if an error is hit, we want it to end, but it may be missing some documents that it hasn’t processed…. so we would to store the error, do the drain and then return the error.

Reply:
nit: Another thought here, as an alternative if you like it, is to have this just return an `interface{}, error` and then the wrapper around this function which is implemented for each of the modules would do the transformation

Reply:
So NATS via the subscribe returns it via a `[]bytes`. So I didn't see a need to return an `interface{}` as this will always be consistent. The unmarshal would happen in each module.

Reply:
Agreed on the error drain. Fixing it now.

Reply:
@lumjjb thoughts on the above comment?